File: linda-core/src/main/java/it/unibo/coordination/linda/core/events/OperationEvent.java
package it.unibo.coordination.linda.core.events;

import it.unibo.coordination.linda.core.*;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class OperationEvent<T extends Tuple, TT extends Template> extends TupleSpaceEvent<T, TT> {

    private final OperationType operationType;
    private final OperationPhase operationPhase;
    private final List<T> argumentTuples;
    private final List<TT> argumentTemplates;
    private final List<T> resultTuples;
    private final List<TT> resultTemplates;

    private OperationEvent(TupleSpace<T, TT> tupleSpace, OperationType operationType, OperationPhase operationPhase,
                           Stream<? extends T> argumentTuples, Stream<? extends TT> argumentTemplates, Stream<? extends T> resultTuples, Stream<? extends TT> resultTemplates) {
        super(tupleSpace);
        this.operationType = Objects.requireNonNull(operationType);
        this.operationPhase = Objects.requireNonNull(operationPhase);
        this.argumentTuples = argumentTuples.collect(Collectors.toList());
        this.argumentTemplates = argumentTemplates.collect(Collectors.toList());
        this.resultTuples = resultTuples.collect(Collectors.toList());
        this.resultTemplates = resultTemplates.collect(Collectors.toList());
    }

    public static <X extends Tuple, Y extends Template> Invocation<X, Y> nothingAcceptingInvocation(TupleSpace<X, Y> tupleSpace, OperationType operationType) {
        if (!OperationType.isNothingAccepting(operationType))
            throw new IllegalArgumentException(operationType.toString());

        return new Invocation<>(
                tupleSpace, operationType, Stream.empty(), Stream.empty()
        );
    }

    public static <X extends Tuple, Y extends Template> Invocation<X, Y> tupleAcceptingInvocation(TupleSpace<X, Y> tupleSpace, OperationType operationType, X tuple) {
        if (!OperationType.isTupleAcceptingSet(operationType))
            throw new IllegalArgumentException(operationType.toString());

        return new Invocation<>(
                tupleSpace, operationType, Stream.of(tuple), Stream.empty()
        );
    }

    public static <X extends Tuple, Y extends Template> Invocation<X, Y> tuplesAcceptingInvocation(TupleSpace<X, Y> tupleSpace, OperationType operationType, Collection<? extends X> tuples) {
        if (!OperationType.isTuplesAcceptingSet(operationType))
            throw new IllegalArgumentException(operationType.toString());

        return new Invocation<>(
                tupleSpace, operationType, tuples.stream(), Stream.empty()
        );
    }

    public static <X extends Tuple, Y extends Template> Invocation<X, Y> templateAcceptingInvocation(TupleSpace<X, Y> tupleSpace, OperationType operationType, Y template) {
        if (!OperationType.isTemplateAccepting(operationType))
            throw new IllegalArgumentException(operationType.toString());

        return new Invocation<>(
                tupleSpace, operationType, Stream.empty(), Stream.of(template)
        );
    }

    public static <X extends Tuple, Y extends Template> Invocation<X, Y> templatesAcceptingInvocation(TupleSpace<X, Y> tupleSpace, OperationType operationType, Collection<? extends Y> templates) {
        if (!OperationType.isTemplatesAccepting(operationType))
            throw new IllegalArgumentException(operationType.toString());

        return new Invocation<>(
                tupleSpace, operationType, Stream.empty(), templates.stream()
        );
    }

    public OperationType getOperationType() {
        return operationType;
    }

    public OperationPhase getOperationPhase() {
        return operationPhase;
    }

    public Optional<T> getArgumentTuple() {
        return argumentTuples.stream().findFirst();
    }

    public Optional<TT> getArgumentTemplate() {
        return argumentTemplates.stream().findFirst();
    }

    public List<T> getArgumentTuples() {
        return List.copyOf(argumentTuples);
    }

    public List<TT> getArgumentTemplates() {
        return List.copyOf(argumentTemplates);
    }

    public boolean isArgumentPresent() {
        return argumentTuples.size() > 0 || argumentTemplates.size() > 0;
    }

    public Optional<T> getResultTuple() {
        return resultTuples.stream().findFirst();
    }

    public Optional<TT> getResultTemplate() {
        return resultTemplates.stream().findFirst();
    }

    public List<T> getResultTuples() {
        return List.copyOf(resultTuples);
    }

    public List<TT> getResultTemplates() {
        return List.copyOf(resultTemplates);
    }

    public boolean isResultPresent() {
        return resultTuples.size() > 0 || resultTemplates.size() > 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        OperationEvent<?, ?> that = (OperationEvent<?, ?>) o;
        return operationType == that.operationType &&
                operationPhase == that.operationPhase &&
                Objects.equals(argumentTuples, that.argumentTuples) &&
                Objects.equals(argumentTemplates, that.argumentTemplates) &&
                Objects.equals(resultTuples, that.resultTuples) &&
                Objects.equals(resultTemplates, that.resultTemplates);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), operationType, operationPhase, argumentTuples, argumentTemplates, resultTuples, resultTemplates);
    }

    @Override
    public String toString() {
        return OperationEvent.class.getSimpleName() + "." + getClass().getSimpleName() + "{" +
                "tupleSpace=" + getTupleSpaceName() +
                ", operationType=" + operationType +
                ", operationPhase=" + operationPhase +
                ", argumentTuples=" + argumentTuples +
                ", argumentTemplates=" + argumentTemplates +
                ", resultTuples=" + resultTuples +
                ", resultTemplates=" + resultTemplates +
                '}';
    }

    public static final class Invocation<T extends Tuple, TT extends Template> extends OperationEvent<T, TT> {

        private Invocation(TupleSpace<T, TT> tupleSpace, OperationType operationType, Stream<? extends T> argumentTuples, Stream<? extends TT> argumentTemplates) {
            super(tupleSpace, operationType, OperationPhase.INVOCATION, argumentTuples, argumentTemplates, Stream.empty(), Stream.empty());
        }

        public Completion<T, TT> toTupleReturningCompletion(T tuple) {
            if (!OperationType.isTupleReturningSet(getOperationType()))
                throw new IllegalStateException();

            return new Completion<>(this, Stream.of(tuple), Stream.empty());
        }

        public Completion<T, TT> toTuplesReturningCompletion(T... tuples) {
            return toTuplesReturningCompletion(Arrays.asList(tuples));
        }

        public Completion<T, TT> toTuplesReturningCompletion(Collection<? extends T> tuples) {
            if (!OperationType.isTuplesReturningSet(getOperationType()))
                throw new IllegalStateException();

            return new Completion<>(this, tuples.stream(), Stream.empty());
        }

        public Completion<T, TT> toTemplateReturningCompletion(TT template) {
            if (!OperationType.isTemplateReturning(getOperationType()))
                throw new IllegalStateException();

            return new Completion<>(this, Stream.empty(), Stream.of(template));
        }

        public Completion<T, TT> toTemplatesReturningCompletion(Collection<? extends TT> templates) {
            if (!OperationType.isTemplatesReturning(getOperationType()))
                throw new IllegalStateException();

            return new Completion<>(this, Stream.empty(), templates.stream());
        }
    }

    public static final class Completion<T extends Tuple, TT extends Template> extends OperationEvent<T, TT> {

        private Completion(Invocation<T, TT> invocation, Stream<? extends T> resultTuples, Stream<? extends TT> resultTemplates) {
            super(
                    invocation.getTupleSpace(),
                    invocation.getOperationType(),
                    OperationPhase.COMPLETION,
                    invocation.getResultTuples().stream(),
                    invocation.getArgumentTemplates().stream(),
                    resultTuples,
                    resultTemplates
            );
        }
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/events/OperationEvent.kt
package it.unibo.coordination.linda.core.events

import it.unibo.coordination.linda.core.OperationPhase
import it.unibo.coordination.linda.core.OperationType
import it.unibo.coordination.linda.core.Template
import it.unibo.coordination.linda.core.Tuple
import it.unibo.coordination.utils.emptyMultiSet
import it.unibo.coordination.utils.multiSetOf
import it.unibo.coordination.utils.toMultiSet
import org.apache.commons.collections4.MultiSet
import java.util.*
import java.util.stream.Stream
import kotlin.streams.toList

sealed class OperationEvent<T : Tuple<T>, TT : Template<T>>(
        override val tupleSpaceName: String,
        open val operationType: OperationType,
        open val operationPhase: OperationPhase,
        open val argumentTuples: List<T> = emptyList(),
        open val argumentTemplates: List<TT> = emptyList(),
        open val resultTuples: MultiSet<T> = emptyMultiSet(),
        open val resultTemplates: MultiSet<TT> = emptyMultiSet(),
        open val results: List<Any> = emptyList()
) : TupleSpaceEvent<T, TT> {

    val argumentTuple: Optional<T>
        get() = argumentTuples.stream().findFirst()

    val argumentTemplate: Optional<TT>
        get() = argumentTemplates.stream().findFirst()

    val isArgumentPresent: Boolean
        get() = argumentTuples.isNotEmpty() || argumentTemplates.isNotEmpty()

    val result: Optional<Any>
        get() = results.stream().findFirst()

    val resultTuple: Optional<T>
        get() = resultTuples.stream().findFirst()

    val resultTemplate: Optional<TT>
        get() = resultTemplates.stream().findFirst()

    val isResultPresent: Boolean
        get() = resultTuples.isNotEmpty() || resultTemplates.isNotEmpty()

    data class Invocation<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val operationType: OperationType,
            override val argumentTuples: List<T>,
            override val argumentTemplates: List<TT>
    ) : OperationEvent<T, TT>(tupleSpaceName,
            operationType,
            OperationPhase.INVOCATION,
            argumentTuples,
            argumentTemplates) {

        private fun toCompletion(
                resultTuples: MultiSet<T> = emptyMultiSet(),
                resultTemplates: MultiSet<TT> = emptyMultiSet(),
                results: List<Any> = emptyList()
        ): Completion<T, TT> {
            return Completion(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTuples = argumentTuples,
                    argumentTemplates = argumentTemplates,
                    resultTuples = resultTuples,
                    resultTemplates = resultTemplates,
                    results = results
            )
        }

        fun toTupleReturningCompletion(tuple: T): Completion<T, TT> {
            check(OperationType.isTupleReturningSet(operationType))

            return toCompletion(
                    resultTuples = multiSetOf(tuple)
            )
        }

        fun toTuplesReturningCompletion(vararg tuples: T): Completion<T, TT> {
            return toTuplesReturningCompletion(Stream.of(*tuples))
        }

        fun toTuplesReturningCompletion(tuples: Stream<out T>): Completion<T, TT> {
            check(OperationType.isTuplesReturningSet(operationType))

            return toCompletion(
                    resultTuples = tuples.map { it }.toMultiSet()
            )
        }

        fun toTuplesReturningCompletion(tuples: Collection<T>): Completion<T, TT> {
            return toTuplesReturningCompletion(tuples.stream())
        }

        fun toTemplateReturningCompletion(template: TT): Completion<T, TT> {
            check(OperationType.isTemplateReturning(operationType))

            return toCompletion(
                    resultTemplates = multiSetOf(template)
            )
        }

        fun toTemplatesReturningCompletion(templates: Collection<TT>): Completion<T, TT> {
            check(OperationType.isTemplatesReturning(operationType))

            return toCompletion(
                    resultTemplates = templates.toMultiSet()
            )
        }

        fun toAnyReturningCompletion(results: Collection<Any>): Completion<T, TT> {
            check(OperationType.isAnyReturning(operationType))

            return toCompletion(
                    results = results.toList()
            )
        }

        fun toCompletion(resultTuples: Stream<out T>, resultTemplates: Stream<out TT>): Completion<T, TT> {
            return toCompletion(
                    resultTuples = resultTuples.map { it }.toMultiSet(),
                    resultTemplates = resultTemplates.map { it }.toMultiSet()
            )
        }
    }

    data class Completion<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val operationType: OperationType,
            override val argumentTuples: List<T>,
            override val argumentTemplates: List<TT>,
            override val resultTuples: MultiSet<T>,
            override val resultTemplates: MultiSet<TT>,
            override val results: List<Any>
    ) : OperationEvent<T, TT>(
            tupleSpaceName,
            operationType,
            OperationPhase.COMPLETION,
            argumentTuples,
            argumentTemplates,
            resultTuples,
            resultTemplates,
            results)

    companion object {

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> of(
                tupleSpaceName: String,
                operationType: OperationType,
                operationPhase: OperationPhase,
                argumentTuples: Stream<out X> = Stream.empty(),
                argumentTemplates: Stream<out Y> = Stream.empty(),
                resultTuples: Stream<out X> = Stream.empty(),
                resultTemplates: Stream<out Y> = Stream.empty(),
                results: Stream<Any> = Stream.empty(),
        ): OperationEvent<X, Y> {
            return when (operationPhase) {
                OperationPhase.INVOCATION -> invocation(
                        tupleSpaceName = tupleSpaceName,
                        operationType = operationType,
                        argumentTuples = argumentTuples,
                        argumentTemplates = argumentTemplates
                )
                OperationPhase.COMPLETION -> completion(
                        tupleSpaceName = tupleSpaceName,
                        operationType = operationType,
                        argumentTuples = argumentTuples,
                        argumentTemplates = argumentTemplates,
                        resultTuples = resultTuples,
                        resultTemplates = resultTemplates,
                        results = results
                )
            }
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> invocation(
                tupleSpaceName: String,
                operationType: OperationType,
                argumentTuples: Stream<out X> = Stream.empty(),
                argumentTemplates: Stream<out Y> = Stream.empty()
        ): Invocation<X, Y> {
            return Invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTuples = argumentTuples.toList(),
                    argumentTemplates = argumentTemplates.toList()
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> completion(
                tupleSpaceName: String,
                operationType: OperationType,
                argumentTuples: Stream<out X> = Stream.empty(),
                argumentTemplates: Stream<out Y> = Stream.empty(),
                resultTuples: Stream<out X> = Stream.empty(),
                resultTemplates: Stream<out Y> = Stream.empty(),
                results: Stream<out Any> = Stream.empty()
        ): Completion<X, Y> {
            return Completion(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTuples = argumentTuples.toList(),
                    argumentTemplates = argumentTemplates.toList(),
                    resultTuples = resultTuples.map { it }.toMultiSet(),
                    resultTemplates = resultTemplates.map { it }.toMultiSet(),
                    results = results.toList()
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> nothingAcceptingInvocation(tupleSpaceName: String, operationType: OperationType): Invocation<X, Y> {
            require(OperationType.isNothingAccepting(operationType)) { operationType.toString() }

            return invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> tupleAcceptingInvocation(tupleSpaceName: String, operationType: OperationType, tuple: X): Invocation<X, Y> {
            require(OperationType.isTupleAcceptingSet(operationType)) { operationType.toString() }

            return invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTuples = Stream.of(tuple)
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> tuplesAcceptingInvocation(tupleSpaceName: String, operationType: OperationType, tuples: Collection<X>): Invocation<X, Y> {
            require(OperationType.isTuplesAcceptingSet(operationType)) { operationType.toString() }

            return invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTuples = tuples.stream()
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> templateAcceptingInvocation(tupleSpaceName: String, operationType: OperationType, template: Y): Invocation<X, Y> {
            require(OperationType.isTemplateAccepting(operationType)) { operationType.toString() }

            return invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTemplates = Stream.of(template)
            )
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> templatesAcceptingInvocation(tupleSpaceName: String, operationType: OperationType, templates: Collection<Y>): Invocation<X, Y> {
            require(OperationType.isTemplatesAccepting(operationType)) { operationType.toString() }

            return invocation(
                    tupleSpaceName = tupleSpaceName,
                    operationType = operationType,
                    argumentTemplates = templates.stream()
            )
        }
    }
}